### PR TITLE
Allow SyGuS subsolver to be reused in incremental mode

### PR DESCRIPTION
**Pull request by ajreynol**
This allows SyGuS subsolvers to be called multiple times to produce solutions for a given set of SyGuS constraints using a new command `check-synth-next`.

By default, the SyGuS subsolver will generate a *new* solution for each successive check-synth.

This simplifies the internal SyGuS solver in several ways for this purpose.  It also ensures that solutions are cached; current master may recompute solutions in the same state more than once if asked, which is no longer the case.

This completes our support for incremental SyGuS.

### CODE REVIEW

File: src/theory/quantifiers/sygus/synth_conjecture.cpp [hunk 1]
```diff
@@ -93,8 +94,22 @@ SynthConjecture::~SynthConjecture() {}
 
 void SynthConjecture::presolve()
 {
-  // we don't have a solution yet
-  d_hasSolution = false;
+  // If we previously had a solution, block it. Notice that
+  // excludeCurrentSolution in the block below ensures we implement a
+  // policy where a *new* solution is generated for check-synth in the set of
```
Comment (HanielB) [inline src/theory/quantifiers/sygus/synth_conjecture.cpp:99]:
```suggestion
  // policy where a *new* solution is generated for check-synth if the set of
```

**Review by HanielB — COMMENTED** [review]
Is there any way we can guard for being running a new `check-synth` command but no different solution existing? So that we could mark that there is no *new* solution rather than no solution?

**Comment by ajreynol** [comment]
Hm maybe we should just do `check-synth-next`....

File: src/smt/solver_engine.cpp [hunk 2]
```diff
@@ -934,11 +934,17 @@ void SolverEngine::assertSygusInvConstraint(Node inv,
...
-Result SolverEngine::checkSynth()
+Result SolverEngine::checkSynth(bool isNext)
 {
   SolverEngineScope smts(this);
   finishInit();
-  return d_sygusSolver->checkSynth(*d_asserts);
+  if (isNext)
+  {
+    
+  }
```
Comment (HanielB) [inline src/smt/solver_engine.cpp:943]: ?

Comment (HanielB) [reply src/smt/solver_engine.cpp]: Ah, sorry, I looked at the wrong version

**Review by HanielB — APPROVED** [review]

**Comment by ajreynol** [comment]
Marking do not merge for now, will need to add a unit test.

File: test/regress/regress1/sygus/incremental-stream-ex.sy [hunk 3]
```diff
@@ -0,0 +1,28 @@
+; COMMAND-LINE: -i --sygus-out=status
+;EXPECT: unsat
+;EXPECT: unsat
+;EXPECT: unsat
+;EXPECT: unsat
```
Comment (abdoo8080) [inline test/regress/regress1/sygus/incremental-stream-ex.sy:5]: This test does not ensure that the results of `check-synth-next` are different. How about adding this one?
```
; COMMAND-LINE: -i
; SCRUBBER: grep "define-fun" | sort
; EXPECT: (define-fun f ((x Int) (y Int)) Int 0)
; EXPECT: (define-fun f ((x Int) (y Int)) Int 1)
; EXPECT: (define-fun f ((x Int) (y Int)) Int x)
; EXPECT: (define-fun f ((x Int) (y Int)) Int y)

(set-logic LIA)

(synth-fun f ((x Int) (y Int)) Int
  ((Start Int))
  ((Start Int (0 1 x y))))

(check-synth)
(check-synth-next)
(check-synth-next)
(check-synth-next)
```
It requires #7819 though...

Comment (ajreynol) [reply test/regress/regress1/sygus/incremental-stream-ex.sy:5]: Sounds good.

Comment (ajreynol) [reply test/regress/regress1/sygus/incremental-stream-ex.sy:5]: I'll add this regression after that PR is merged.

**Pull request merged**